feat: recover payments on failed days

Tiavina-Andriamamivony opened the pull request.

Tiavina-Andriamamivony committed:
feat(recovery): add service and endpoint for Orange transaction recovery

Added OrangeTransactionRecoveryService to fetch Orange transactions for a given date, insert missing transactions, and trigger corresponding payment verifications via PaymentVerificationRequestedService. Added OrangeRecoveryController with POST /api/orange/recover?date=YYYY-MM-DD endpoint to trigger recovery and return a JSON summary. Enhanced logging for fetch, insert, and payment verification steps. Added unit tests: OrangeTransactionRecoveryServiceTest (Mockito) for service logic and OrangeRecoveryControllerTest (MockMvc) for endpoint behavior. Notes: endpoint is idempotent and should be secured before exposure.

Tiavina-Andriamamivony committed:
feat: find payments feature

Tiavina-Andriamamivony committed:
chore: format

Tiavina-Andriamamivony committed:
chore: replace env conf data

Tiavina-Andriamamivony committed:
chore(test): replace test data

Tiavina-Andriamamivony committed:
chore: format

Tiavina-Andriamamivony committed:
Merge branch 'hei-school:preprod' into preprod

Tiavina-Andriamamivony committed:
Merge branch 'verifyAll' into preprod

Tiavina-Andriamamivony committed:
chore: delete comments

Tiavina-Andriamamivony committed:
chore: clean

Tiavina-Andriamamivony committed:
fix: multiple payment verification

Delete custom jrepository method, return failed payments

Tiavina-Andriamamivony committed:
chore: format

Tiavina-Andriamamivony committed:
test: expect FAILED payments for non-existing payment infos

Tiavina-Andriamamivony committed:
chore(recovery): set endpoint type to put and replace response entity

Tiavina-Andriamamivony committed:
chore(orange): simplify transaction recovery service and improve code readability

- Remove inner RecoveryResult class in favor of sync.model.RecoveryResult
- Extract methods to follow Single Responsibility Principle
- Add meaningful method names
- Improve error handling with granular try-catch blocks
- Separate concerns: persistence, verification, and result building
- Update tests to use new RecoveryResult model and sync() method
- Add test cases for failure and partial success scenarios

Tiavina-Andriamamivony committed:
chore: format

Tiavina-Andriamamivony committed:
chore: refactor PaymentRepository for better readability

- Extract findPaymentsByPaymentInfos logic into atomic methods (findOrCreateFailedPayment, mapToFoundPayment, createFailedPlaceholderPayment)
- Replace magic number 6 with FAILED_PAYMENT_ATTEMPT_COUNT constant
- Use Optional.map/orElseGet instead of verbose if/else
- Add [PAYMENT_SEARCH] log prefix for consistency
- Add Javadoc to document placeholder FAILED payment behavior
- Improve testability through separation of concerns

Tiavina-Andriamamivony committed:
chore: delete excessive log and add some tests

Tiavina-Andriamamivony committed:
chore: refactor Orange sync service and simplify tests

Tiavina-Andriamamivony committed:
chore: revert randomEmail result

Tiavina-Andriamamivony committed:
chore: refactor Orange sync service and simplify tests

Tiavina-Andriamamivony committed:
test: remove redundant OrangeRecoveryTest

Tiavina-Andriamamivony committed:
chore(orangeSync): refactor

Tiavina-Andriamamivony committed:
refactor: remove batch payment verification from sync PR

Tiavina-Andriamamivony committed:
chore: restore PaymentServiceIT

Tiavina-Andriamamivony committed:
chore: compact formatting

Tiavina-Andriamamivony committed:
chore: refactor payment repository

Tiavina-Andriamamivony committed:
chore: format

Tiavina-Andriamamivony committed:
test: keep only essential sync test

Tiavina-Andriamamivony committed:
chore: revert payment repository

Tiavina-Andriamamivony committed:
chore: format

Tiavina-Andriamamivony committed:
chore: delete line

hei-teacher commented on src/main/java/school/hei/vola/endpoint/rest/controller/PaymentController.java:
```diff
@@ -35,4 +41,10 @@ public Payment getPayment(
+
+  @PutMapping("/orange/sync")
+  public RecoveryResult sync(
+      @RequestParam("date") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate date) {
```
just ISO.DATE is enough

hei-teacher commented on src/main/java/school/hei/vola/service/sync/model/RecoveryResult.java:
```diff
@@ -0,0 +1,8 @@
+
+@Builder
+public record RecoveryResult(
+    LocalDate date, boolean isSuccessful, int inserted, String errorMessage) {}
```
inserted... inserted what?

For example, recoveredTransactionsNb is VASTLY SUPERIOR (read with a posh accent)

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,72 @@
+      var inserted = (int) transactions.stream().filter(this::processTransaction).count();
+      return RecoveryResult.builder().date(date).isSuccessful(true).inserted(inserted).build();
+    } catch (Exception e) {
+      log.error("[SYNC] Failed date={}", date, e);
```
log.info as you already return appropriately

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,72 @@
+    }
+  }
+
+  private boolean processTransaction(OrangeTransaction tx) {
```
process what...

Use a verb that explicitly tell what you are trying to do

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,72 @@
+      triggerVerificationIfExists(tx);
+      return inserted;
+    } catch (Exception e) {
+      log.error("[SYNC] Failed ref={}", tx.getRef(), e);
```
use log.info as you already return properly

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  private boolean persistIfNew(OrangeTransaction tx) throws JsonProcessingException {
+    if (transactionRepository.existsById(tx.getRef())) return false;
```
never remove on one-instruction if

Tiavina-Andriamamivony committed:
chore: refactor

Use ISO.date and static import, clarify process method name, use log.info instead of log.error

hei-teacher commented on src/main/java/school/hei/vola/service/sync/model/RecoveryResult.java:
```diff
@@ -0,0 +1,8 @@
+
+@Builder
+public record RecoveryResult(
+    LocalDate date, boolean isSuccessful, int transactionsInserted, String errorMessage) {}
```
nbTranstractionsInserted

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,78 @@
+  public RecoveryResult sync(LocalDate date) {
+    try {
+      var transactions = orangeApiClient.transactionsOf(date).getTransactions();
+      var inserted = (int) transactions.stream().filter(this::persistAndVerifyTransaction).count();
```
(int) is called and unsafe cast. Avoir unsafe as much as possible.
Do ::mapToInt then ::sum for example

More importantly, your return type is finally NOT appropriate.
Cuz you just return the number of successfully inserted transactions.
But you do not tell how many of them SHOULD be sucessful.

ie, RecoveryResult= ..., nbSyncedTransactions, nbTotalTransactions, ...

Then, use log.error again instead of log.info __BUT__ putting the failed transactions details in the error so we can see in Sentry which are precisely the transactions that could be successfully synced.

Tiavina-Andriamamivony committed:
refactor: improve RecoveryResult and error logging

- Replace unsafe cast with mapToInt().sum()
- Add nbSyncedTransactions and nbTotalTransactions to track success rate
- Use log.error with transaction details for Sentry visibility
- Include ref, amount, and full transaction object in error logs

Tiavina-Andriamamivony committed:
chore: add comment

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,92 @@
+
+      return RecoveryResult.builder()
+          .date(date)
+          .isSuccessful(true)
```
isSuccessful(syncedTransactions == totalTransactions)

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,92 @@
+      return inserted;
+    } catch (Exception e) {
+      log.error(
+          "[SYNC] Failed to sync transaction ref={}, amount={}, details={}",
```
"[SYNC] Failed to sync orange transaction ot={}, tx" is vastly superior, with appropriate ::toString on OrangeTransaction.

Furthermore ot instead of tx is better.

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,92 @@
+    if (transactionRepository.existsById(tx.getRef())) {
+      return false;
+    }
+    var entity = new JOrangeTransaction();
```
No JXxx things should be directly visible in core domain.

You should only have access to TransactionRepository here, __not__ JTransactionRepository

Tiavina-Andriamamivony committed:
chore: add to string to orange transaction

Tiavina-Andriamamivony committed:
feat: save method to orange payment repository

Tiavina-Andriamamivony committed:
chore: update orange sync service to use save of orange sync service

Tiavina-Andriamamivony committed:
chore: format

hei-teacher commented on src/main/java/school/hei/vola/service/OrangeSyncService.java:
```diff
@@ -0,0 +1,82 @@
+    try {
+      boolean inserted = persistIfNew(ot);
+      triggerVerificationIfExists(ot);
+      return inserted;
```
Hm... Nah... It could happen that some transactions of the day were successfully persisted prior to the day. In such a case nbSynced != nbTotal hence you will have isSuccessful == false.

That would be problematic since your sync is actually succfessful, in the sense that you only persisted what you needed to persist.

You should always return true if you reach here. And make persistIfNew return void.

And and a new test to exhibit that specific case.

Tiavina-Andriamamivony committed:
fix: sync success should not depend on insert count

Tiavina-Andriamamivony committed:
chore: test data
